### DISCUSSION TITLE
How to use signing with nFPM?

### DISCUSSION THREAD

**Discussion by SSSOCPaulCote** [goreleaser/nfpm#563] [Q&A] [answered]
I'm very new when it comes to software releases and I found nFPM which has made things very easy. The next step in the learning process for me is signing binaries and release packages. I followed [this tutorial](https://www.digitalocean.com/community/tutorials/how-to-use-reprepro-for-a-secure-package-repository-on-ubuntu-14-04) on how to run my own APT repo and generate my own signing keys.

TL;DR of the tutorial: using gpg, I generated a master key and a subkey for signing from it. Exported the master key to a file and backed that up on a separate machine. Deleted the file and the master key from the keyring, which then deleted all other keys. Imported the signing key. 

Now when I write my configuration files for nFPM, I try to export the signing key into a file format as follows:
```
$ gpg --export --armor XXXXXXXX > key.gpg
```
This prompts me for my passphrase. In the config under `deb`, `rpm` and `apk` I put this block:
```yaml
signature:
    key_file: key.gpg
    key_id: XXXXXXXX
```
When I go to create my `.deb`, `.rpm` and `.apk` packages, I get this:
```bash
using deb packager...
signing error: armored detach sign: no signing key in keyring
using rpm packager...
failed to create signatures: call to signer failed: signing error: no signing key in keyring
using apk packager...
signing error: no PEM block found
```
I've tried exporting the passphrase manually as the `NFPM_PASSPHRASE` variable and using `env NFPM_PASSPHRASE=` as a prefix to the `nfpm` command, no luck. Can someone guide me here?

**Comment by caarlos0** [comment, marked as answer]
as stated in the docs, the env variable has a different name: https://goreleaser.com/customization/nfpm/

in your case, probably `NFPM_DEFAULT_PASSPHRASE`, unless you set a different `nfpms.id`

----

###### please reply in thread instead of a new one, thank

**Reply by SSSOCPaulCote** [reply]
Hello, thank you for responding. To be clear, I'm not using goreleaser and just using nfpm. I will try that ENV variable though.

**Reply by caarlos0** [reply]
then you should have opened the issue in the [nfpm repo](https://github.com/goreleaser/nfpm/discussions) (EDIT: just transferred it).

that said, follow the instruction in the [nfpm docs](https://nfpm.goreleaser.com)

**Reply by SSSOCPaulCote** [reply]
When I click Issues, then New Issue on the nfpm github, I am greeted with this menu
![image](https://user-images.githubusercontent.com/89483361/196709783-a2c14f39-5e7d-4898-9fdc-1fbdecbdae3c.png)
And I clicked Ask a question, which brought me to the goreleaser discussion. I assumed nfpm was fair game, my bad.
Thank you for transferring it.

The [nfpm docs](https://nfpm.goreleaser.com/configuration/) mention that the ENV variable is either `NFPM_RPM_PASSPHRASE` with a fallback of `NFPM_PASSPHRASE`, both of which I have tried. I have not tried the `deb` and `apk` equivalents though. I'll give that a shot.

**Reply by caarlos0** [reply]
oh, my bad then, just fixed that link in fc0fe55

**Reply by SSSOCPaulCote** [reply]
I tried using `NFPM_DEB_PASSPHRASE`, `NFPM_RPM_PASSPHRASE`, `NFPM_APK_PASSPHRASE` and just `NFPM_PASSPHRASE` as per docs and I keep getting the same errors.
```
using deb packager...
signing error: armored detach sign: no signing key in keyring
using rpm packager...
failed to create signatures: call to signer failed: signing error: no signing key in keyring
using apk packager...
signing error: no PEM block found
```
when I do `gpg --list-keys` I see the key I'm referencing. I exported the key in this way `gpg --export --armor XXXXXXXX > key.gpg`

**Reply by caarlos0** [reply]
it looks there are no keys in the `key.gpg` file, or the key can't be used for signing... or its just the public key.

https://github.com/goreleaser/nfpm/blob/fc0fe55cf73eb029a88b480ba3217910f9a4691d/internal/sign/pgp.go#L230 

`errNoKeys` is the error you are getting... so the password in the env is not the issue it seems.

**Reply by SSSOCPaulCote** [reply]
Hey I figured it out, you're right I was writing the public key to the file instead of the private key. `gpg --export-secret-key --armor XXXXXXXX > key.gpg`. Thanks!
